L70-node-proto-loader-type-generator.md

```diff
@@ -0,0 +1,237 @@
+@grpc/proto-loader TypeScript Type Generator CLI Tool
+----
+* Author(s): murgatroid99
+* Approver: wenbozhu
+* Status: Draft
+* Implemented in: Node.js
+* Last updated: 2020-06-01
+* Discussion at: https://groups.google.com/g/grpc-io/c/FwLprMC-5Qc
+
+## Abstract
+
+Add a tool to the `@grpc/proto-loader` to generate types that describe the objects that will be generated by loading the output of `@grpc/proto-loader` into `grpc` or `@grpc/grpc-js`.
+
+## Background
+
+`@grpc/proto-loader` outputs objects with types determined by the `.proto` files loaded at runtime, so it is currently very difficult to get compile-time type information about those objects. As TypeScript becomes increasingly popular, that compile-time type information becomes increasingly desirable.
+
+
+### Related Proposals: 
+* L23 Standalone Node.js gRPC+Protobuf.js Library API
+* L43 Node Message Type Information
+
+## Proposal
+
+In the `@grpc/proto-loader` library, provide a command line tool that will generate the types of the objects that will be created by passing the output of `load` or `loadSync` to `grpc.loadPackageDefinition` when loading a specific set of `.proto` files with a specific set of options. This will allow users to write code like this to use the type information in the rest of the code:
+
+```ts
+import * as grpc from 'grpc';
+// OR 
+import * as grpc from '@grpc/grpc-js';
+
+import { ProtoGrpcType, ServiceHandlers, messages } from 'path/to/proto-file-name_proto'; // File generated by the tool is path/to/proto-file-name_proto.d.ts
+import * as protoLoader from '@grpc/proto-loader';
+
+const packageDefinition = protoLoader.loadSync('other/path/to/proto-file-name.proto', options);
+const loadedPackageDefinition = grpc.loadPackageDefinition(packageDefinition) as ProtoGrpcType; // How the types will be used
+
+const requestObject: messages.fully.qualified.package.MessageName = {
+  field1: value1,
+  field2: value2
+};
+
+const serviceHandler: ServiceHandlers.fully.qualified.package.ServiceName = {
+  // Call implicitly has the correct arity (unary, client streaming, etc.) and the call and callback have the correct request and response types
+  methodName(call, callback) {
+    // Implementation
+  }
+}
+
+const server = new grpc.Server();
+server.addService(loadedPackageDefinition.fully.qualified.package.ServiceName.service, serviceHandler);
+
+```
+
+The tool will be named `proto-loader-gen-types` and will have this usage:

```

Comment:
Will this work as a protoc plugin? Or will it stand on it's own.

Reply:
This is a standalone tool, distributed in the `@grpc/proto-loader` package. It uses Protobuf.js rather than protoc for parsing.